Support 0-batch size for nn.Linear.

kennyhorror opened the pull request:
Summary:
At the current moment of time nn.Linear (an it's interal functional code), will
fail in THBlas:

RuntimeError: invalid argument 8: lda should be at least max(1, 0), but have 0 at caffe2/aten/src/TH/generic/THBlas.cpp:363

This diff is trying to fix this bug.

As of now I was able to identify 2 possible places where changes needs to be done based on current dispatcher logic:
1. The file touched in this diff
2. caffe2/aten/src/THC/generic/THCTensorMathBlas.cu

At the moment I didn't find a better places comparing to injecting logic to those files:
the only non-generated function for forward pass, this + mm_mat2_backward function family on a backward pass.

Test Plan: New unit-tests are passing. Code that was failing earlier works. Need to test other backends.

Differential Revision: D17599915

SsnL commented on aten/src/TH/generic/THTensorMath.cpp:
```diff
@@ -298,20 +298,44 @@ static void THTensor_(addmmImpl)(THTensor *r_, THTensor *t, THTensor *m1, THTens
-                ldr__);
+  // Don't go through GEMM if result is empty matrix, since this is not
+  // supported by BLAS.
+  if (m != 0 && n != 0) {
```
I'm a bit confused... If the result is empty, what are these assignments doing here?

kennyhorror replied:
There're 2 main cases when failure in BLAS routines happens:

1. n or m is 0, which will result in the final matrix that 0 as one of the dimensions (as an example Linear called on a batch of size 0). In this case the only part that is needed from this function - resize of the final matrix to have correct output dimensions.

2. Another case when BLAS can fail is when k is 0, that can fail when one of the strides is 0 (happens on backward pass for Linear with an empty batch). In that case I'm getting rid of no-op part for ADDMM:
alpha * A x B + beta * C -> beta * C

dzhulgakov commented on aten/src/TH/generic/THTensorMath.cpp:
```diff
@@ -298,20 +298,44 @@ static void THTensor_(addmmImpl)(THTensor *r_, THTensor *t, THTensor *m1, THTens
+          }
+        }
+      }
+      else {
```
usually beta is 1.0 so we can skip this loop all together

kennyhorror committed:
Support 0-batch size for nn.Linear. (#27211)

Summary:
At the current moment of time nn.Linear (an it's interal functional code), will
fail in THBlas:

RuntimeError: invalid argument 8: lda should be at least max(1, 0), but have 0 at caffe2/aten/src/TH/generic/THBlas.cpp:363

This diff is trying to fix this bug.

As of now I was able to identify 2 possible places where changes needs to be done based on current dispatcher logic:
1. The file touched in this diff
2. caffe2/aten/src/THC/generic/THCTensorMathBlas.cu

At the moment I didn't find a better places comparing to injecting logic to those files:
the only non-generated function for forward pass, this + mm_mat2_backward function family on a backward pass.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27211

Test Plan: New unit-tests are passing. Code that was failing earlier works. Need to test other backends.

Differential Revision: D17599915

Pulled By: kennyhorror

fbshipit-source-id: cd47fe9fddb3bad1be5ddaddb1c1d9b95a76d258

ezyang commented:
This seems to break py2.7.9 test

```

Oct 09 00:47:41 test_interface (__main__.TestClassType) ... terminate called after throwing an instance of 'std::runtime_error'
Oct 09 00:47:41   what():  pybind11_object_dealloc(): Tried to deallocate unregistered instance!
Oct 09 00:47:41 Traceback (most recent call last):
Oct 09 00:47:41   File "test/run_test.py", line 458, in <module>
Oct 09 00:47:41     main()
Oct 09 00:47:41   File "test/run_test.py", line 450, in main
Oct 09 00:47:41     raise RuntimeError(message)
Oct 09 00:47:41 RuntimeError: test_jit failed! Received signal: SIGIOT
```
